Mathlib/Data/Bitvec/Defs.lean

```diff
@@ -1,339 +1,177 @@
+instance : Div (BitVec w) := inferInstanceAs (Div (Fin _))
+instance : LT (BitVec w)  := ⟨fun x y => BitVec.lt x y⟩
+instance : LE (BitVec w)  := ⟨fun x y => BitVec.le x y⟩
 
-instance {n} : ShiftLeft (Bitvec n) := ⟨fun x y => x <<< y.toNat⟩
+/-!
+## Bitwise operations
+-/
 
-instance {n} : ShiftRight (Bitvec n) := ⟨fun x y => x >>> y.toNat⟩
+/-- The complement of a bitvector. -/
+protected def complement (x : BitVec w) : BitVec w :=
+  0 - (x + .ofNat w 1)
+
+/-- The and of two bitvectors. -/
+protected def and (x y : BitVec w) : BitVec w :=
+  ⟨x.val &&& y.val, by simp [HAnd.hAnd, AndOp.and, Nat.land, Nat.bitwise_lt]⟩
+
+/-- The or of two bitvectors. -/
+protected def or (x y : BitVec w) : BitVec w :=
+  ⟨x.val ||| y.val, by simp [HOr.hOr, OrOp.or, Nat.lor, Nat.bitwise_lt]⟩
+
+/-- The xor of two bitvectors. -/
+protected def xor (x y : BitVec w) : BitVec w :=
+  ⟨x.val ^^^ y.val, by simp [HXor.hXor, Xor.xor, Nat.xor, Nat.bitwise_lt]⟩
+
+/-- The left-shift of a bitvector by some amount. -/
+protected def shiftLeft (x : BitVec w) (n : Nat) : BitVec w :=
+  .ofNat w (x.val <<< n)
+
+/-- The right-shift of a bitvector by some amount. -/
+protected def shiftRight (x : BitVec w) (n : Nat) : BitVec w :=
+  ⟨x.val >>> n, by
+      simp only [Nat.shiftRight_eq_div_pow]
+      exact lt_of_le_of_lt (Nat.div_le_self' _ _) (x.isLt) ⟩
+
+/-- Signed less than for bitvectors. -/
+protected def slt (x y : BitVec (w + 1)) : Prop :=
+  ((x.val >>> w = 1) ∧ (y.val >>> w = 0)) ∨ ((x.val >>> w = y.val >>> w) ∧ x < y)
+
+/-- Signed less than or equal to for bitvectors. -/
+protected def sle (x y : BitVec (w + 1)) : Prop :=
+  ((x.val >>> w = 1) ∧ (y.val >>> w = 0)) ∨ ((x.val >>> w = y.val >>> w) ∧ x ≤ y)
+
+/-- Signed greater than for bitvectors. -/
+protected def sgt (x y : BitVec (w + 1)) : Prop := BitVec.slt y x
+
+/-- Signed greater than or equal to for bitvectors. -/
+protected def sge (x y : BitVec (w + 1)) : Prop := BitVec.sle y x
+
+instance : Complement (BitVec w) := ⟨BitVec.complement⟩
+instance : AndOp (BitVec w) := ⟨BitVec.and⟩
+instance : OrOp (BitVec w) := ⟨BitVec.or⟩
+instance : Xor (BitVec w) := ⟨BitVec.xor⟩
+instance : HShiftLeft (BitVec w) Nat (BitVec w) := ⟨BitVec.shiftLeft⟩
+instance : HShiftRight (BitVec w) Nat (BitVec w) := ⟨BitVec.shiftRight⟩
+
+/-- Left rotation by some amount. -/
+def rotateLeft (x : BitVec w) (n : Nat) : BitVec w :=
+  x <<< n ||| x >>> (w - n)
+
+/-- Right rotation by some amount. -/
+def rotateRight (x : BitVec w) (n : Nat) : BitVec w :=
+  x >>> n ||| x <<< (w - n)
+
+/-- Concatenation of two bitvectors. -/
+protected def append (x : BitVec w) (y : BitVec v) : BitVec (w + v) :=
+  ⟨x.val <<< v ||| y.val, Nat.add_comm _ _ ▸ Nat.append_lt y.isLt x.isLt⟩
+
+instance : HAppend (BitVec w) (BitVec v) (BitVec (w + v)) := ⟨BitVec.append⟩
+
+/-- Extract the bitvector between indices i and j. -/
+@[simp]
+def extract (i j) (x : BitVec w) : BitVec (i - j + 1) :=
+  BitVec.ofNat _ (x.val >>> j)
```

Comment:
I would argue this should be
```suggestion
@[simp]
def extract (i j) (x : BitVec w) : BitVec (i - j) :=
  BitVec.ofNat _ (x.val >>> j)
```
that is, `extract 0 8 x` obtains the lowest 8 bits, and `extract i j x ++ extract j k x = extract i k x` (or perhaps the other way around, since you didn't document `++` clearly enough!)